File: 3234-count-the-number-of-substrings-with-dominant-ones/3234-count-the-number-of-substrings-with-dominant-ones.py
class Solution(object):
    def numberOfSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        v = deque()
        for i in range(n):
            if s[i] == '0':
                v.append(i)
        ans = 0
        for i in range(n):
            while v and v[0] < i:
                v.popleft()
            prev = i - 1
            cnt1 = 0
            cnt0 = 0
            for x in v:
                if x - prev - 1 >= max(0, cnt0 * cnt0 - cnt1):
                    ans += x - prev - 1 - max(0, cnt0 * cnt0 - cnt1)
                    if max(0, cnt0 * cnt0 - cnt1) != 0:
                        ans += 1
                cnt1 += x - prev - 1
                prev = x
                cnt0 += 1
                if cnt0 * cnt0 <= cnt1:
                    ans += 1
                if cnt0 * cnt0 > cnt1 + n - x - 1:
                    break
                prev = x
            if n - prev - 1 >= max(0, cnt0 * cnt0 - cnt1):
                ans += n - prev - 1 - max(0, cnt0 * cnt0 - cnt1)
                if max(0, cnt0 * cnt0 - cnt1) != 0:
                    ans += 1
        return ans

File: 3234-count-the-number-of-substrings-with-dominant-ones/3234-count-the-number-of-substrings-with-dominant-ones.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int ans = 0, n = s.size();
        vector<int> pref(n);
        for (int i = 0; i < n; ++i) {
            pref[i] = s[i] == '1';
            if (i) pref[i] += pref[i-1];
        }
        
        vector<int> zeros;
        for (int i = 0; i < n; ++i) {
            if (s[i] == '0') zeros.push_back(i);
            
            for (int k = 1; k <= size(zeros); ++k) {
                if (k*k >= i+2) break;
                
                int R = zeros[size(zeros) - k];
                int L = -1;
                if (k < size(zeros)) L = zeros[size(zeros) - k - 1];
                int lo = 0;
                if (L >= 0) lo = pref[L];
                
                int hi = pref[R];
                ans += max(0, min(hi, pref[i] - k*k) - lo + 1);
            }
            
            if (s[i] == '1') {
                if (size(zeros) == 0) ans += i+1;
                else ans += i - zeros.back();
            }
        }
        return ans;
    }
};